driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -23,12 +34,87 @@
 public class OperationContext {
     private static final AtomicLong NEXT_ID = new AtomicLong(0);
     private final long id;
+    private final ServerDeprioritization serverDeprioritization;
 
     public OperationContext() {
         id = NEXT_ID.incrementAndGet();
+        serverDeprioritization = new ServerDeprioritization();
     }
 
     public long getId() {
         return id;
     }
+
+    /**
+     * @return The same {@link ServerDeprioritization} if called on the same {@link OperationContext}.
+     */
+    public ServerDeprioritization getServerDeprioritization() {
+        return serverDeprioritization;
+    }
+
+    public static final class ServerDeprioritization {
+        @Nullable
+        private ServerAddress candidate;
+        private final Set<ServerAddress> deprioritized;
+
+        private ServerDeprioritization() {
+            candidate = null;
+            deprioritized = new HashSet<>();
+        }
+
+        ServerSelector apply(final ServerSelector selector) {
+            return new DeprioritizingSelector(selector);
```

Comment:
This `apply` naming is confusing: I expected it to perhaps apply certain settings to the selector, but actually it wraps in another selector.

It seems the purpose of CompositeServerSelector to handle such "chaining", so I would expect a `DeprioritizingSelector` to be another entry in the list in `getCompositeServerSelector`.

Reply:
> It seems the purpose of CompositeServerSelector to handle such "chaining"

`CompositeServerSelector` just applies one selector after the other, and that's not the logic needed for the new functionality.

Reply:
I thought more about the possibility to implement server selection via a `DeprioritizingSelector` that is composed using `CompositeServerSelector` either before or after `settings.getServerSelector()` (regardless of whether that order is immediate). As far as I can see, that approach can't work: see counterexamples below.

`DeprioritizingSelector.select(clusterDescription)` may act in one of the three ways:

1. Filter: return a `List<ServerDescription>` that does not contain deprioritized servers, provided that there are normal servers; otherwise return everything.
2. Mark: return a `List<ServerDescription>` that contains essentially the same elements as `clusterDescription.getServerDescriptions`, with the only difference that some `ServerDescription`s are marked with the priority. This requires changing the API of the `ServerDescription`. This fact makes this approach unfavourable.
3. Sort: return a `List<ServerDescription>` that contains the same elements as `clusterDescription.getServerDescriptions`, but ordered according to the priority. Unfortunately, the order of elements in `ServerDescription.getServerDescriptions` does not seem to have any meaning, as `ServerDescription.equals` treats two `ServerDescription`s with different orders of elements in `getServerDescriptions` as equal. This means, we should not rely on the order.

Now let's consider all possibilities

# `DeprioritizingSelector` is before `settings.getServerSelector()`

## 1. Filter

- Consider `settings.getServerSelector()` that selects only servers with the `t` tag.
- `clusterDescription.getServerDescriptions` contains descriptions with server addresses `a`, `b`, only `b` has the `t` tag, only `b` is deprioritized.
- `DeprioritizingSelector` filters `b` out, and `settings.getServerSelector()` has only `a` to select from.
- `settings.getServerSelector()` selects nothing, as `a` does not have the `t` tag.
- According to the spec, server selection should have also considered `b` and selected despite it being deprioritized, because without considering this server, selection fails. However, there is no way to consider `b` in this scenario, and we failed the server selection.

## 2. Mark

- Consider `settings.getServerSelector()` that selects one server at random.
- `clusterDescription.getServerDescriptions` contains descriptions with server addresses `a`, `b`, only `b` is deprioritized.
- `DeprioritizingSelector` marks `b`'s `ServerDescription` as deprioritized, but `settings.getServerSelector()` knows nothing about such marks, and operates as if nothing were marked.
- `settings.getServerSelector()` selects `b` at random, despite it being deprioritized.
- We should have selected `a`, but we incorrectly selected the deprioritized server `b`.

## 3. Sort

The counterexample here is essentially identical to "2. Mark", as `settings.getServerSelector()` knows nothing about the order of `clusterDescription.getServerDescriptions` representing priorities. Actually, even if it new about that, it still could not have worked correctly all the time, as it needs to be able to clearly discriminate between deprioritized and normal servers when selecting a server at random.

# `settings.getServerSelector()` is before `DeprioritizingSelector`

## 1. Filter

- Consider `settings.getServerSelector()` that selects one server at random.
- `clusterDescription.getServerDescriptions` contains descriptions with server addresses `a`, `b`, only `b` is deprioritized.
- `settings.getServerSelector()` selects `b` at random, despite it being deprioritized.
- `DeprioritizingSelector` selects `b` as it has no normal servers to select.
- We should have selected `a`, but we incorrectly selected the deprioritized server `b`.

## 2. Mark

The counterexample here is essentially identical to "1. Filter", as `DeprioritizingSelector` can't do anything for `a` to be selected.

## 3. Sort

The counterexample here is essentially identical to "1. Filter", as `DeprioritizingSelector` can't do anything for `a` to be selected.

P.S. All of the counterexamples above work with the original approach taken in this PR. However, when we talked, we also identified a bug when combining this approach with the way `BaseCluster.selectServer(ServerSelector, ClusterDescription, Function<ServerAddress, Server> serverCatalog)` deals with the internal race condition (that code was introduced in https://github.com/mongodb/mongo-java-driver/commit/f19447753657d7be3a597e31731f5aedf0cc9ccd). That can be dealt with by addressing the internal race condition differently: instead of calling `getServer` and checking if the result is `null`, we can take a non-linearizable snapshot of the `Map` that `getServer` uses internally, filter the `ClusterDescription` using that snapshot, and forget about any troubles.

Reply:
The [spec](https://github.com/mongodb/specifications/blob/d3a355c299d9840bf9a758910a1739fcbc33ec64/source/server-selection/server-selection.md?plain=1#L673) provides an algorithm in which deprioritization (filtering out of deprioritized servers, when others are available) occurs in step 3 and completes before step 4 (our `settings.getServerSelector`). This PR does not adhere to the spec: it implements a different algorithm, which "wraps" steps 3-7 and repeats them when no servers are found (for any reason).

I believe this is a spec bug. If the existing latency selector comes after the user selector, then so should the deprioritizing selector. "Optimizing" selectors must be applied after any selector expressing necessary conditions, because all requirements must be met before something is selected due to preference.

The spec should be changed (to correct this, and for clarity), to list the following sequence of composed selectors:
```
serverSelector
settings.getServerSelector (ask user)
filter by latency
randomize
deprioritize stable sort
2 first
pick 1 by least ops
```
I see no issue with relying on sorting/ordering in our internal code (that is, all selectors that follow the user-provided selector).

Reply:
Implemented the deprioritization logic according to the spec in https://github.com/mongodb/mongo-java-driver/pull/1367/commits/b3430bd0d60838fdeff00ef682857279b56d0747.